module/VuFind/src/VuFind/ILS/Driver/Alma.php

```diff
@@ -233,114 +263,121 @@ protected function getAvailabilityFromItem($item)
-                            );
-
-                            // Add all allowed request types to an array
-                            $requestTypesArr = [];
-                            foreach ($requestTypes as $requestType) {
-                                $requestTypesArr[] = (string)$requestType;
-                            }
-
-                            // If HOLD is an allowed request type, add the link for
-                            // placing a hold
-                            $addLink = in_array('HOLD', $requestTypesArr);
-                        }
-
-                        $results[] = [
-                            'id' => $id,
-                            'source' => 'Solr',
-                            'availability' => $this->getAvailabilityFromItem($item),
-                            'status' => (string)$item
-                                ->item_data
-                                ->base_status[0]
-                                ->attributes()['desc'],
-                            'location' => $locationCode,
-                            'reserve' => 'N',   // TODO: support reserve status
-                            'callnumber' => (string)$item->holding_data->call_number,
-                            'duedate' => $duedate,
-                            'returnDate' => false, // TODO: support recent returns
-                            'number' => $number,//++$copyCount,
-                            'barcode' => empty($barcode) ? 'n/a' : $barcode,
-                            'item_notes' => $itemNotes,
-                            'item_id' => $itemId,
-                            'holding_id' => $holdingId,
-                            'addLink' => $addLink ?? false,
-                               // For Alma title-level hold requests
-                            'description' => $description
-                        ];
+        $itemLimit = $this->getItemLimit();
+        $page = $page ?? 1;
+        $offset = ($page * $itemLimit) - $itemLimit;
+
+        // The path for the API call. We call "ALL" available items, but not at once.
+        // The "limit" tells the API how many items should be returned by the call
+        // at once (e. g. 10). The "offset" defines the range (e. g. get items 30 to
+        // 40). With these parameters we are able to use a paginator for paging
+        // through many items.
+        $itemsPath = '/bibs/' . urlencode($id) . '/holdings/ALL/items?limit='
+            . $itemLimit . '&offset=' . $offset
+            . '&order_by=library,location,enum_a,enum_b&direction=desc';
+
+        if ($items = $this->makeRequest($itemsPath)) {
+            // Get the total number of items returned from the API call and set it to
+            // a class variable. It is then used in VuFind\RecordTab\HoldingsILS for
+            // the items paginator.
+            $this->totalItemCount = (int)$items->attributes()->total_record_count;
```

Comment:
Similar to the above comment, I'm not a big fan of introducing a new property and method just to pass this total value back to the calling code. This might actually be a situation where we should consider changing the return format of `getHolding` to better reflect pagination. Instead of just being an array of holdings arrays, we could start returning a top-level array with 'total' and 'holdings' keys. The VuFind\ILS\Connection class could normalize this by checking for the existince of a 'total' key in the data returned by the driver, and building the data structure as needed with a `count` call for non-paginated lists. Obviously that's a somewhat disruptive change, but as long as we support the existing drivers in a backward-compatible way, it won't be terribly intrusive, and I think it will give us a cleaner interface from the higher-level code.

Reply:
OK, I will try to change the return array as you suggested. A question concerning not being disruptive to other ILS drivers: Is it save to test the changes against the Demo driver to look for errors? I think this is the only option I have for being sure that I don't break other drivers as I don't have access to another ILS than Alma.

Reply:
The `totalItemCount`property I introduced here is removed now. As you suggested I changed the return value of `getHolding()`. Now, a top level array with "total" and "holdings" keys is returned. In "ILS\Connection" in the method `getHolding()` the array is normalized if a "total" key exists so that subsequent code can process it. But I found no other solution than re-introducing the property `totalItemCount` that I removed from "ILS\Driver\Alma" there. That is because I must be able to get the value in "RecordTab\HoldingsILS" for the paginator so it knows the total number of items it has to page through. `Count`ing the items returned from the API would not be possible here because the API (at least the one from Alma, but possibly also from other ILS) limits the returned items to 100, but there could be e. g. 342 items. So the paginator must know that it has to page through 342 items and not only 100. But I think your goal was to put the code at a level where other ILS drivers could profit from it and I think that should work now. I tested against the "Demo" driver and no problems occured.